### PR TITLE
EPMDJ-7281 Fixed multi-instances attach, refactoring

### PR DESCRIPTION
**Pull request by ViktoriiaLebedeva**

### CODE REVIEW

File: core/src/main/kotlin/agent/Extensions.kt [hunk 1]
```diff
@@ -67,15 +59,16 @@ internal fun AgentInfo.toDto(
     agentManager: AgentManager,
 ): AgentInfoDto = run {
     val plugins = agentManager.plugins.ofAgent(this)
-    val instanceIds = agentManager.instanceIds(id).keys
+    val instanceIds = agentManager.instanceIds(id).keys.toSet()
```
Comment (Mazel-Tovr) [inline core/src/main/kotlin/agent/Extensions.kt]: why u do keys.toSet ? 
```suggestion
    val instanceIds = agentManager.instanceIds(id).keys
```

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 2]
```diff
@@ -208,37 +223,37 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
+            logger.info { "Instance '${instanceId}' of Agent '${id}' was disconnected" }
         }
     }
 
     internal fun instanceIds(
         agentId: String,
-    ): PersistentMap<String, InstanceState> = getOrNull(agentId)?.let { agentInfo ->
-        agentInfo.instances.also {
-            logger.trace { "instances ids of agent(id=$agentId, version=${agentInfo.buildVersion}): ${it.keys}" }
+    ) = entryOrNull(agentId)?.info?.run {
```
Comment (ViktoriiaLebedeva) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: use getOrNull()

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 3]
```diff
@@ -208,37 +223,37 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
+        val key = AgentKey(agentId, buildVersion)
+        _instances.value[key]?.takeIf { it.isNotEmpty() }?.also {
+            logger.trace { "instances ids of agent(id=$agentId, version=$buildVersion): ${it.keys}" }
```
Comment (ViktoriiaLebedeva) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: remove this

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 4]
```diff
@@ -527,10 +541,10 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
         agentId: String,
         instanceId: String,
         block: suspend AgentWsSession.() -> Unit,
-    ): Unit = entryOrNull(agentId)?.run {
-        getInstanceState(instanceId)?.let { instanceState ->
-            val id = info.id
-            updateInstanceStatus(instanceId, AgentStatus.BUSY).also { instance ->
+    ): Unit = entryOrNull(agentId)?.info?.run {
```
Comment (ViktoriiaLebedeva) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: use getOrNull()

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 5]
```diff
@@ -540,13 +554,43 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
+                updateInstanceStatus(instanceKey, instanceId, AgentStatus.ONLINE).also {
                     notifyAgents(id)
                     logger.debug { "Agent $id is online." }
                 }
             }
         } ?: logger.warn { "Instance $instanceId is not found" }
     } ?: logger.warn { "Agent $agentId not found" }
+
+    internal fun updateInstanceStatus(
+        instanceKey: AgentKey,
```
Comment (ViktoriiaLebedeva) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: rename

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 6]
```diff
@@ -540,13 +554,43 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
+        _instances.updateAndGet { instances ->
+            logger.trace { "instance $instanceId changed status, new status is $status" }
+            instances[instanceKey]?.let {
+                instances.put(instanceKey, it.put(instanceId, instanceState.copy(status = status)))
+            } ?: instances
+        }[instanceKey]
+    } ?: persistentMapOf()
+
+    private fun getInstanceState(
+        instanceKey: AgentKey,
```
Comment (ViktoriiaLebedeva) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: rename

File: core/src/main/kotlin/endpoints/AgentManager.kt [hunk 7]
```diff
@@ -540,13 +554,43 @@ class AgentManager(override val kodein: Kodein) : KodeinAware {
...
+        instanceId: String
+    ) = _instances.value[instanceKey]?.get(instanceId)
+
+    fun getStatus(agentId: String): AgentStatus = instanceIds(agentId).let { instances ->
+        if (entryOrNull(agentId)?.info?.isRegistered == true) {
```
Comment (ViktoriiaLebedeva) [inline core/src/main/kotlin/endpoints/AgentManager.kt]: use getOrNull()

**Pull request merged**